Command module: add server-side argument types

Leo40Git opened the pull request:
Basically an almost-direct port of the Colonel API (https://gitlab.com/stellardrift/colonel), minus syncing stuff (which requires networking, and such depends on QuiltMC#34).

Leo40Git committed:
Command module: add server-side argument types

Basically an almost-direct port of the Colonel API (https://gitlab.com/stellardrift/colonel), minus syncing stuff (which requires networking, and such depends on QuiltMC#34).

Leo40Git committed:
Add Colonel licenses to all files sourced from Colonel

Leo40Git committed:
Add enum argument type

This functions as a test of the ServerArgumentType API, and is also pretty convenient, IMO
The serializer leaves much to be desired, though...

Leo40Git committed:
EnumArgumentType: rewrite to make less horrible

Leo40Git committed:
EnumArgumentType: add translation for "unknown value" exception

Leo40Git committed:
EnumArgumentType: add sanity checks

OroArmor commented:
Networking is merged, so syncing can be implemented.

Leo40Git committed:
Merge branch '1.18' into server-arg-types

Leo40Git committed:
Implement known argument type syncing

Leo40Git committed:
license

also add missing ApiStatus.Internal annotations

Leo40Git commented:
Syncing added! Note that Quilt Networking is an _optional_ dependency - if it isn't present, the server will assume all clients don't know about the extra argument types.

OroArmor commented on library/command/command/src/main/java/org/quiltmc/qsl/command/impl/KnownArgumentTypesSync.java:
```diff
@@ -0,0 +1,61 @@
+import org.quiltmc.qsl.networking.api.client.C2SPlayChannelEvents;
+
+@ApiStatus.Internal
+public final class KnownArgumentTypesSync {
```
This can probably be done with an entry point event, removing the extra `if (loaded) ...`

Leo40Git replied:
Oh, right, forgot about Aurora's new innovations.

Leo40Git replied:
Actually, wait, the server side can't be done with entrypoints since it isn't actually an event.

Leo40Git replied:
![image](https://user-images.githubusercontent.com/7997354/152642569-9106a2cd-d16a-4eb6-bc83-6abc3f3fde14.png)

OroArmor commented on library/command/command/build.gradle:
```diff
@@ -10,4 +10,12 @@ qslModule {
+
+dependencies {
+	// OPTIONAL dependency on networking
+	compileOnly project(path: ":core:networking", configuration: "dev")
```
CC @QuiltMC/qsl-core-library should we use compile only or just impl? i can update #72 to reflect the decision here

TheGlitch76 commented:
> Syncing added! Note that Quilt Networking is an optional dependency - if it isn't present, the server will assume all clients don't know about the extra argument types.

Has this been tested to work?

Leo40Git commented:
No, but it's pretty much a direct copy-paste from Colonel - so if it doesn't work I'd be surprised.

TheGlitch76 commented on library/command/command/src/main/java/org/quiltmc/qsl/command/impl/ClientInitializer.java:
```diff
@@ -0,0 +1,35 @@
+public final class ClientInitializer implements ClientModInitializer {
+	@Override
+	public void onInitializeClient() {
+		if (FabricLoader.getInstance().isModLoaded("quilt_networking")) {
```
Should log if support is enabled/disabled

TheGlitch76 commented on library/command/command/src/main/java/org/quiltmc/qsl/command/impl/Initializer.java:
```diff
@@ -0,0 +1,48 @@
+
+	@Override
+	public void onInitialize() {
+		if (FabricLoader.getInstance().isModLoaded("quilt_networking")) {
```
Should log if support is enabled/disabled

TheGlitch76 commented on library/command/command/src/main/java/org/quiltmc/qsl/command/impl/PreLaunchHacks.java:
```diff
@@ -0,0 +1,65 @@
+ * Original on GitHub at <a href="https://github.com/i509VCB/Fabric-Junkkyard/blob/ce278daa93804697c745a51af06ec812896ec2ad/src/main/java/me/i509/junkkyard/hacks/PreLaunchHacks.java">i509VCB/Fabric-Junkkyard</a>
+ */
+@ApiStatus.Internal
+public final class PreLaunchHacks {
```
Since this could have unintended side effects (mixing into Brigader works sometimes, but not all the time), maybe this would be better as a `core` module?

Leo40Git replied:
This was already moved to `core`.

LambdAurora commented on library/command/command/build.gradle:
```diff
@@ -7,6 +7,15 @@ qslModule {
 	version = "1.0.0"
 	library = "command"
 	coreDependencies([
```
This should be updated to reflect [the new dependency declaration](https://github.com/QuiltMC/quilt-standard-libraries/blob/1.18/CONTRIBUTING.md#declaring-dependencies-between-modules)

LambdAurora commented on library/command/command/src/main/java/org/quiltmc/qsl/command/api/EnumArgumentType.java:
```diff
@@ -0,0 +1,137 @@
+import net.minecraft.server.command.ServerCommandSource;
+import net.minecraft.text.TranslatableText;
+
+public final class EnumArgumentType implements ArgumentType<String> {
```
A javadoc on the class would be welcomed.

Leo40Git replied:
I tried

LambdAurora commented on library/command/command/src/main/java/org/quiltmc/qsl/command/api/ServerArgumentType.java:
```diff
@@ -0,0 +1,114 @@
+	/**
+	 * Gets the identifier of this argument type.
+	 *
+	 * @return argument type ID
```
```suggestion
	 * @return argument type identifier
```

LambdAurora commented on library/command/command/src/main/java/org/quiltmc/qsl/command/impl/ClientInitializer.java:
```diff
@@ -0,0 +1,35 @@
+
+@ApiStatus.Internal
+@Environment(EnvType.CLIENT)
+public final class ClientInitializer implements ClientModInitializer {
```
This should be updated to use the new QSL's `ClientModInitializer`.

LambdAurora commented on library/command/command/src/main/java/org/quiltmc/qsl/command/impl/Initializer.java:
```diff
@@ -0,0 +1,48 @@
+import org.quiltmc.qsl.command.api.ServerArgumentType;
+
+@ApiStatus.Internal
+public final class Initializer implements ModInitializer {
```
This should be updated to use the new QSL's `ModInitializer`.

LambdAurora commented on library/command/command/src/main/java/org/quiltmc/qsl/command/impl/PreLaunchHacks.java:
```diff
@@ -0,0 +1,65 @@
+ * Original on GitHub at <a href="https://github.com/i509VCB/Fabric-Junkkyard/blob/ce278daa93804697c745a51af06ec812896ec2ad/src/main/java/me/i509/junkkyard/hacks/PreLaunchHacks.java">i509VCB/Fabric-Junkkyard</a>
+ */
+@ApiStatus.Internal
+public final class PreLaunchHacks {
```
This probably should be put in QSL base, in the impl package?
So if another module needs it for some reason it has access to it, but it should stay as impl and not API.

LambdAurora commented on library/command/command/src/main/java/org/quiltmc/qsl/command/impl/ServerArgumentTypeImpl.java:
```diff
@@ -0,0 +1,44 @@
+
+@ApiStatus.Internal
+public record ServerArgumentTypeImpl<T extends ArgumentType<?>>(Identifier id, Class<? extends T> type,
+																ArgumentSerializer<T> serializer,
```
Extremely nitpicky, but maybe this should use smart tabs?
In GitHub I see it with a 2 character indentation-tab, and it's quite unaligned.

LambdAurora commented on library/command/command/src/main/java/org/quiltmc/qsl/command/mixin/CommandManagerMixin.java:
```diff
@@ -44,4 +62,29 @@ void addQuiltCommands(CommandManager.RegistrationEnvironment environment, Callba
+		// Repeat as long as a type is replaceable -- that way you can have a hierarchy of argument types.
+		while (type != null && !knownExtraCommands.contains(type.id())) {
+			((RequiredArgumentBuilderAccessor) builder).setType(type.fallbackProvider().createVanillaFallback(builder.getType()));
+			if (type.fallbackSuggestions() != null) {
```
In QSL we usually try to keep blank lines around new blocks for readability.

Leo40Git replied:
Forgot to run checkstyle for this PR...

LambdAurora reviewed (approved):
This needs to be updated since 22w06a happened.

Once that's done, it's globally good!

LambdAurora committed:
Merge upstream.

Leo40Git committed:
ID -> identifier

Co-authored-by: LambdAurora <aurora42lambda@gmail.com>

Leo40Git committed:
EnumArgumentType: add javadoc

add EnumArgumentType.getEnum to match other ArgumentType classes

Leo40Git committed:
ServerArgumentTypeImpl: use spaces for record definition indentation

there, now it doesn't look godawful on GitHub

LambdAurora committed:
Move PreLaunchHacks, add server arg testmod, fix up some issues.

Oliver-makes-code commented:
Do the arguments also work on the integrated server?

LambdAurora commented:
> Do the arguments also work on the integrated server?

Yes, and one of the premises is if possible, it'll use the actual argument type on the client if present using synchronization code too.

TheGlitch76 commented on library/command/command/src/main/java/org/quiltmc/qsl/command/api/EnumArgumentType.java:
```diff
@@ -0,0 +1,190 @@
+/**
+ * An {@link ArgumentType} that allows an arbitrary set of (case-insensitive) strings.
+ *
+ * <p>Using this argument type will <em>not</em> prevent Vanilla clients from joining your server.
```
```suggestion
 * <p>This argument type is compatible with Vanilla clients.
```

TheGlitch76 commented on library/command/command/src/main/java/org/quiltmc/qsl/command/api/EnumArgumentType.java:
```diff
@@ -0,0 +1,190 @@
+		return StringArgumentType.StringType.SINGLE_WORD.getExamples();
+	}
+
+	public static final class Serializer implements ArgumentSerializer<EnumArgumentType> {
+		@Override
+		public void toPacket(EnumArgumentType type, PacketByteBuf buf) {
+			buf.writeCollection(type.values, PacketByteBuf::writeString);
+		}
+
+		@Override
+		public EnumArgumentType fromPacket(PacketByteBuf buf) {
+			Set<String> values = buf.readCollection(LinkedHashSet::new, PacketByteBuf::readString);
+			return new EnumArgumentType(values);
+		}
+
+		@Override
+		public void toJson(EnumArgumentType type, JsonObject json) {
+			var valuesArr = new JsonArray();
+
+			for (var value : type.values) {
+				valuesArr.add(value);
+			}
+
+			json.add("values", valuesArr);
+		}
+	}
```
Enums with the same value names aren't necessarily the same enum. This should keep track of the target class and only allow serialization to the correct class IMO.

Leo40Git replied:
`EnumArgumentType` isn't enum-specific, though - it specifically handles a value set.
`enumConstant` and `getEnumConstant` are simply utility methods to use arbitrary enum classes as arguments. ~~Though, maybe I should add a notice that they're "unchecked" in their javadocs...~~

Leo40Git replied:
`enumConstant` and `getEnumConstant` now have their own "is this the correct enum class" checks. As a fun side-effect, `enumConstant` now caches return values.

TheGlitch76 commented on library/command/command/src/main/java/org/quiltmc/qsl/command/impl/ServerArgumentTypes.java:
```diff
@@ -0,0 +1,70 @@
+		if (player instanceof ServerPlayerEntity serverPlayer) {
+			((ServerPlayerEntityHooks) serverPlayer).quilt$setKnownArgumentTypes(types);
+			if (!types.isEmpty()) {
+				// TODO avoid resending the whole command tree, find a way to receive the packet before sending?
```
did you mean to implement this?

Leo40Git replied:
I should look into this at some point, but this TODO is copied from Colonel.

Leo40Git replied:
...well, I implemented this, but I also broke the feature entirely, so...

Leo40Git replied:
I figured out how I broke the feature! So uh, resolving this.

TheGlitch76 commented on library/command/command/src/main/java/org/quiltmc/qsl/command/mixin/CommandManagerMixin.java:
```diff
@@ -46,11 +61,41 @@
 		return a == b || b == CommandManager.RegistrationEnvironment.ALL;
 	}
+
+	// region Copyright 2016, 2017, 2018, 2019, 2020 zml and Colonel contributors
```
```suggestion
	// region Copyright 2016, 2017, 2018, 2019, 2020 zml and Colonel contributors.
	// Licensed under the Apache 2.0 License; see this file's header for additional information.
```

TheGlitch76 commented on library/core/qsl_base/src/main/java/org/quiltmc/qsl/base/impl/PreLaunchHacks.java:
```diff
@@ -0,0 +1,67 @@
+ */
+@ApiStatus.Experimental
+@ApiStatus.Internal
+public final class PreLaunchHacks {
```
Definitely would rather avoid putting this into QSL. We will be switching to quilt-loader very soon, perhaps we could implement something there to replace this?
cc @AlexIIL

AlexIIL replied:
Allowing mixins / chasm to target libraries other than mods and minecraft seems reasonable.

AlexIIL replied:
Mixins can target DFU and Brigadier as of https://github.com/QuiltMC/quilt-loader/commit/8c577ff48653b61178e0f8affeaeff5e8e93c722 - so this PreLaunchHook can probably be removed when QSL targets quilt-loader directly.

Leo40Git committed:
EnumArgumentType: unclunkify vanilla support notice

Co-authored-by: Glitch <glitch.g3431@gmail.com>

Leo40Git committed:
CommandManagerMixin: add details to Colonel section

Co-authored-by: Glitch <glitch.g3431@gmail.com>

Leo40Git committed:
PreLaunchInitializer: add note to remove once QLoader is targeted

Leo40Git committed:
Initializers: log networking support status

Leo40Git committed:
EnumArgumentType: getEnumConstant: verify enum class

enumConstant: cache returned arg types

Leo40Git committed:
Merge branch '1.18' into server-arg-types

# Conflicts:
#	library/command/command/src/main/resources/fabric.mod.json
#	library/core/networking/src/main/resources/fabric.mod.json

Leo40Git committed:
ServerArgumentTypes: don't resend commands unless known argument types have changed

Leo40Git committed:
[NOT WORKING] Sync known argument types earlier to avoid resending command tree

...yeah I somehow broke it
I have no idea how or why, but Vanilla clients can't connect now because they end up parsing a bad identifier somewhere

KnownArgTypesSync: Vanilla clients now get empty set instead of set being null

LambdAurora committed:
Merge upstream.

Leo40Git committed:
[IT WORKS NOW] restore FMJ changes

this indirectly caused Vanilla clients to not be able to connect...
because when Minecraft fails when serializing an argument type, *it sends an empty identifier*, which the client promptly chokes on.
(MC fails to serialize an argument type if it's not registered - and the code that registers EnumArgumentType wasn't running because
I accidentally ran over the FMJ file.)

Great job, Mojang.

LambdAurora committed:
Fix getEnumConstant.

williambl commented on library/command/command/src/main/java/org/quiltmc/qsl/command/impl/ServerArgumentTypeImpl.java:
```diff
@@ -0,0 +1,44 @@
+import org.quiltmc.qsl.command.api.ServerArgumentType;
+
+@ApiStatus.Internal
+public record ServerArgumentTypeImpl<T extends ArgumentType<?>>(Identifier id, Class<? extends T> type,
+                                                                ArgumentSerializer<T> serializer,
+                                                                ArgumentTypeFallbackProvider<T> fallbackProvider,
+                                                                @Nullable SuggestionProvider<?> fallbackSuggestions)
+		implements ServerArgumentType<T> {
```
```suggestion
public record ServerArgumentTypeImpl<T extends ArgumentType<?>>(
		Identifier id,
		Class<? extends T> type,
		ArgumentSerializer<T> serializer,
		ArgumentTypeFallbackProvider<T> fallbackProvider,
		@Nullable SuggestionProvider<?> fallbackSuggestions
) implements ServerArgumentType<T> {
```

IMO this looks nicer, but it doesn't matter all that much.

williambl commented on library/command/command/src/main/java/org/quiltmc/qsl/command/impl/ServerArgumentTypeImpl.java:
```diff
@@ -0,0 +1,44 @@
+import org.quiltmc.qsl.command.api.ServerArgumentType;
+
+@ApiStatus.Internal
+public record ServerArgumentTypeImpl<T extends ArgumentType<?>>(Identifier id, Class<? extends T> type,
+                                                                ArgumentSerializer<T> serializer,
+                                                                ArgumentTypeFallbackProvider<T> fallbackProvider,
+                                                                @Nullable SuggestionProvider<?> fallbackSuggestions)
+		implements ServerArgumentType<T> {
```
```suggestion
public record ServerArgumentTypeImpl<T extends ArgumentType<?>>(
		Identifier id,
		Class<? extends T> type,
		ArgumentSerializer<T> serializer,
		ArgumentTypeFallbackProvider<T> fallbackProvider,
		@Nullable SuggestionProvider<?> fallbackSuggestions
) implements ServerArgumentType<T> {
```

IMO this looks nicer, but it doesn't matter all that much.

Leo40Git replied:
meh, AFAIK the current formatting is the standard?

williambl replied:
fairs.

LambdAurora replied:
> meh, AFAIK the current formatting is the standard?

is it?
I don't think there's a clear standard, I find the second more readable, especially on small-width devices.

williambl commented on library/command/command/src/main/java/org/quiltmc/qsl/command/impl/KnownArgTypesStorage.java:
```diff
@@ -0,0 +1,31 @@
+import net.minecraft.util.Identifier;
+
+@ApiStatus.Internal
+public interface KnownArgTypesStorage {
+	Set<Identifier> quilt$getKnownArgumentTypes();
+
+	void quilt$setKnownArgumentTypes(Set<Identifier> types);
+}
```
I think this can be interface-injected, would make the usages of it cleaner

Leo40Git replied:
Interface injections is meant for API interfaces, not implementation interfaces (AFAIK, anyway).

williambl replied:
True, interface injection would expose this implementation detail. I hadn't considered that.

williambl commented on library/command/command/src/main/java/org/quiltmc/qsl/command/mixin/CommandManagerMixin.java:
```diff
@@ -46,11 +61,42 @@
+	// region Copyright 2016, 2017, 2018, 2019, 2020 zml and Colonel contributors.
+	// Licensed under the Apache 2.0 License; see this file's header for additional information.
+	@SuppressWarnings({"rawtypes", "unchecked"}) // argument type generics
+	@Inject(method = "makeTreeForSource", locals = LocalCapture.CAPTURE_FAILEXCEPTION,
```
Why `FAILEXCEPTION` and not `FAILHARD`?

Leo40Git replied:
No idea! Should I change it to `FAILHARD`?

williambl replied:
`FAILHARD` is used in other parts of QSL, it's probably best to be consistent (and to fail early)

williambl reviewed (commented):
I'd be happy to approve with these questions answered, sorry it took so long for me to get round to reviewing

Leo40Git committed:
Reformat ServerArgumentTypeImpl

Co-authored-by: William Bradford Larcombe <william@williambl.com>

Leo40Git committed:
Make local capture fail hard
